### PR TITLE
JSON parse failures should be 4xx codes

### LINKED ISSUE #112296: IllegalStateExceptions caused by JsonParseExceptions should return 4xx status codes
### Elasticsearch Version

Latest main

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

Serverless and cloud

### Problem Description

Invalid Json that results in an `IllegalStateException` when parsing return 500 status codes and should return 4xx status codes as this indicates invalid output. 

This issue is similar to https://github.com/elastic/elasticsearch/issues/111542 but slightly different as that issue is due to an EOF exception. 

### Steps to Reproduce

Use the following command to reproduce: 

```
curl localhost:9200/_search -H "Content-Type: application/json" -d '{ "query": { "match": { "foo": [ { "bar": { "baz": 1 }} ] }}'
```

The curl response is: 
```
{
  "error": {
    "root_cause": [
      {
        "type": "illegal_state_exception",
        "reason": "Can't get text on a START_ARRAY at 1:32"
      }
    ],
    "type": "illegal_state_exception",
    "reason": "Can't get text on a START_ARRAY at 1:32"
  },
  "status": 500
}
```

### Logs (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by benwtrent**
It seemed if there wasn't any text to parse, this is not an internal issue but instead an argument issue.

I simply changed the exception thrown. If we don't agree with this, I can adjust `query` parsing directly, but this seemed like the better choice.

closes: https://github.com/elastic/elasticsearch/issues/112296

### CODE REVIEW

**Review by ldematte — COMMENTED** [review]
We discussed with @cbuescher whether IllegalArgumentException was a good choice on https://github.com/elastic/elasticsearch/pull/112605. I'm fine with either IllegalArgumentException or ValidationException, but I think that whatever we choose  it's better to be consistent. Wdyt?

File: libs/x-content/impl/src/main/java/org/elasticsearch/xcontent/provider/json/JsonXContentParser.java [hunk 1]
```diff
@@ -111,7 +111,7 @@ public String text() throws IOException {
     }
 
     private void throwOnNoText() {
-        throw new IllegalStateException("Can't get text on a " + currentToken() + " at " + getTokenLocation());
+        throw new IllegalArgumentException("Can't get text on a " + currentToken() + " at " + getTokenLocation());
```
Comment (rjernst) [inline libs/x-content/impl/src/main/java/org/elasticsearch/xcontent/provider/json/JsonXContentParser.java:114]: nit: I think the message could be improved slightly. Maybe something like "Expected text at LOCATION but found TOKEN"

**Review by rjernst — APPROVED** [review]
Looks good, I have one suggestion to make the error message easier to understand

**Comment by benwtrent** [comment]
@ldematte @cbuescher I read the discussion, IAE or an xcontent parsing exception makes sense to me here as we are parsing values input values. 

The PR linked has to do with the content type not even being known. That to me is a horse of a different color. While, so, I would prefer an IAE there as well as the input is the REST request we are handling and thus an illegal argument was given, the poorly formed request.

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Pull request merged**